### PR TITLE
Bug fixes, test file catalog, refactoring

### PR DESCRIPTION
**Pull request by dputtick**
* Symlinks that were linked in a loop could cause an infinite recursion -> fixed.
* Added a harness for testing individual files with an expected outcome (dangerous vs normal).
* Removed the "unknown" and "binary" categories. As we discussed, they were confusing and could all just be considered dangerous.
* Improved weaker spots in FileBase -> added @properties, made things easier to reason about, fixed a few edge case bugs.
* Some of the older kittengroomer unit tests were brittle -> refactored them to use mocks more frequently.

### CODE REVIEW

File: kittengroomer/helpers.py [hunk 1]
```diff
@@ -43,82 +30,44 @@ def __init__(self, src_path, dst_path):
...
-            'errors': {},
-            'user_defined': {}
-        }
-        self.extension = self._determine_extension()
-        self.set_property('extension', self.extension)
-        self.mimetype = self._determine_mimetype()
+        self.dst_dir = os.path.dirname(dst_path)
+        self.filename = os.path.basename(src_path)
+        self.size = self._get_size(src_path)
+        self.is_dangerous = False  # Should this be some other value initially?
```
Comment (Rafiot) [inline kittengroomer/helpers.py]: For binary values, I'd keep it to True or False. Would it make sense to put it to True by default?

File: kittengroomer/helpers.py [hunk 2]
```diff
@@ -236,52 +179,92 @@ def safe_copy(self, src=None, dst=None):
...
         try:
-            dst_path, filename = os.path.split(dst)
-            if not os.path.exists(dst_path):
-                os.makedirs(dst_path)
+            # TODO: maybe don't check and just mkdir? Is there a way to do this that won't raise an Exception? Can we make shutil automatically make dirs inbetween?
+            if not os.path.exists(self.dst_dir):
+                os.makedirs(self.dst_dir)
             shutil.copy(src, dst)
         except Exception as e:
+            # TODO: what exceptions are we expecting to catch here?
```
Comment (Rafiot) [inline kittengroomer/helpers.py]: IOError, mostly in case we are unable to write on the dest key.

File: kittengroomer/helpers.py [hunk 3]
```diff
@@ -236,52 +179,92 @@ def safe_copy(self, src=None, dst=None):
...
-            self.dst_path += ext
-        if not self._file_props['extension'] == ext:
-            self.set_property('extension', ext)
-
-    def create_metadata_file(self, ext):
+    def force_ext(self, extension):
+        """If dst_path does not end in `extension`, append .ext to it."""
+        new_ext = self._check_leading_dot(extension)
+        if not self.filename.endswith(new_ext):
+            # TODO: do we want to log that the extension was changed?
```
Comment (Rafiot) [inline kittengroomer/helpers.py]: I'd say yes.

File: kittengroomer/helpers.py (hunk 1 above)
Comment (dputtick) [inline kittengroomer/helpers.py]: Yeah, I'm not sure. Are we more comfortable assuming that files are dangerous by default or safe?
* If we assume they are safe, we are at risk of programmer error (in not adding a make_dangerous call) or a dependency breaking leading to a false negative.
* If we assume they are dangerous, we have a higher chance of false positives, and we have to add a bunch of cases for files that are safe to mark them as dangerous. This feels worse to me, because it would make the code much more complex and hard to reason about.

We're probably better off assuming safe by default and then having a bunch of test files that we are constantly checking against

File: kittengroomer/helpers.py [hunk 4]
```diff
@@ -236,52 +179,92 @@ def safe_copy(self, src=None, dst=None):
         if dst is None:
             dst = self.dst_path
         try:
-            dst_path, filename = os.path.split(dst)
-            if not os.path.exists(dst_path):
-                os.makedirs(dst_path)
+            # TODO: maybe don't check and just mkdir? Is there a way to do this that won't raise an Exception? Can we make shutil automatically make dirs inbetween?
```
Comment (dputtick) [inline kittengroomer/helpers.py]: One thing to note, the docs for makedirs say that it can still raise an error with an existing directory if the existing dir has different permissions. This can't happen in 3.4.1 or above. I'll make the change, but maybe we should wrap it in its own try/except?

File: bin/filecheck.py [hunk 5]
```diff
@@ -308,13 +324,12 @@ def text(self):
...
-            if subtype in self.sub_type:
-                # TODO: should we change the logic so we don't iterate through all of the subtype methods?
-                # TODO: should these methods return a value?
-                method()
-                return
-        self._unknown_app()
+        if self.subtype in self.app_subtype_methods:
+            method = self.app_subtype_methods[self.subtype]
+            method()
+            # TODO: should these application methods return a value?
```
Comment (dputtick) [inline bin/filecheck.py]: I've been thinking about making the error handling better, this is a good place do it. But I don't think we need a return value. I was thinking about making the core loop of the program more functional, but it doesn't make sense for what we're doing (since we almost always take no action after reaching the end of the call graph, other than copying).

File: bin/filecheck.py [hunk 6]
```diff
@@ -654,8 +682,8 @@ def process_file(self, file):
         if file.should_copy:
             file.safe_copy()
             file.set_property('copied', True)
-        file.write_log()
-        if file.is_recursive:
+            file.write_log()
+        if file.is_archive:
             self.process_archive(file)
```
Comment (dputtick) [inline bin/filecheck.py]: Nope, this is one of the ugly things about processing archives. I really need to clean up this code. The write_log happens in process_archive, because the file needs to be unpacked first. It might be worth it to switch to a workflow where we identify files that are archives before doing anything else, and then have a completely different workflow for them, rather than calling file.check() on every file and then handling archives separately later.

**Comment by dputtick** [comment]
Ok, added comments above and pushed the changes you mentioned

File: bin/filecheck.py (hunk 5 above)
Comment (Rafiot) [inline bin/filecheck.py]: That's right.

File: bin/filecheck.py (hunk 6 above)
Comment (Rafiot) [inline bin/filecheck.py]: That could make the code simpler, yeah. But for now, it can stay as it is.

**Comment by Rafiot** [comment]
Thanks!
